nova/virt/vmwareapi/volumeops.py

```diff
@@ -590,3 +591,74 @@ def attach_root_volume(self, connection_info, instance,
+
+    def fixup_shadow_vms(self, instance, shadow_vms):
+        # This should sensibly moved out to cinder
+        vm_ref = vm_util.get_vm_ref(self._session, instance)
+        for device in vm_util.get_hardware_devices(self._session, vm_ref):
+            class_name = device.__class__.__name__
+
+            if class_name != "VirtualDisk":
+                continue
+
+            try:
+                current_device_path = device.backing.fileName
+                volume = shadow_vms[device.backing.uuid]["volume"]
+                volume_ref = vutil.get_moref(volume, "VirtualMachine")
+                original_device = self._get_vmdk_base_volume_device(volume_ref)
+                if original_device:
+                    original_device_path = original_device.backing.fileName
+                    if original_device_path == current_device_path:
+                        # Already attached
+                        continue
+                    else:
+                        # That should not happen
+                        LOG.warning("Shadow-vm %s has already disk"
+                            " attached at {} replacing it with {}",
+                            volume, original_device_path, current_device_path
+                            )
+                        self.detach_disk_from_vm(self, volume_ref, instance,
+                            original_device, destroy_disk=True)
+
+                disk_type = vm_util._get_device_disk_type(device)
+                self.attach_disk_to_vm(volume_ref,
+                    instance,
+                    constants.DEFAULT_ADAPTER_TYPE,
+                    disk_type,
+                    current_device_path
+                    )
+            except KeyError:
+                # Unknown volume, so we ignore it
+                continue
+            except Exception:
+                LOG.exception("Failed to attach volume {}".format(device.key))
+
+    def delete_shadow_vms(self, block_device_info, instance=None):
+        # We need to delete the migrated shadow vms
+        # (until we implement it in cinder)
+        block_device_mapping = driver.block_device_info_get_mapping(
+            block_device_info)
+
+        if not block_device_mapping:
+            return
+
+        session = self._session
+        deleted = []
+        for disk in block_device_mapping:
+            connection_info = disk["connection_info"]
+            try:
+                data = connection_info["data"]
+                volume_ref = self._get_volume_ref(data["volume"])
+                destroy_task = session._call_method(session.vim,
+                                            "Destroy_Task",
+                                            volume_ref)
+                session._wait_for_task(destroy_task)
+                deleted.append("{volume_id} ({volume})".format(**data))
+            except Exception as e:
+                LOG.error("Failed to delete %(volume)s. "
```

Comment:
The way I see it, `LOG.exception` is for cases, where the exception cannot be handled properly.
We print the whole backtrace to find out where it happens, so we can dig into it an fix it.
I do `LOG.error` because if we failed to destroy the VM, the only thing we can reasonably do is log it.
Well, admittedly `Exception` is quite broad, but if I miss to catch here anything, then we end up with a failed migration and orphan shadow-vms.

See file deletion in  [image cache](https://github.com/openstack/nova/blob/370e322d7393d64abcc5c8da5cb723f886787364/nova/virt/vmwareapi/imagecache.py#L63-L74)  or in [VMops](https://github.com/openstack/nova/blob/6edf06dbfd902f7af413d2d7a93760d05efab4fa/nova/virt/vmwareapi/vmops.py#L191-L201) as precedents.

Reply:
My problem with logging the exception converted to string is, that we sometimes lose info for debugging and thus fixing the problem to not reoccur. Best example: `KeyError` converted to string is only the key, that was missing:
```python
>>> print("{}".format(KeyError('asdf')))
'asdf'
```

Reply:
Okay, I'll log the exception. It should practically never happen (hopefully) anyway. If things are happening more often, we can catch them explicitly.